go/tools/gazelle/merger/merger.go

```diff
@@ -0,0 +1,232 @@
+You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package merger provides methods for merging parsed BUILD files.
+package merger
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"sort"
+
+	bzl "github.com/bazelbuild/buildifier/core"
+)
+
+var (
+	mergeableFields = map[string]bool{
+		"srcs": true,
+		"deps": true,
+	}
+)
+
+// MergeWithExisting looks for an existing BUILD file at file.Path
+// loads it, and attempts to merge elements of newfile into it.
+// returns newfile, nil if FileNotExists
+func MergeWithExisting(newfile *bzl.File) (*bzl.File, error) {
+	b, err := ioutil.ReadFile(newfile.Path)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return newfile, nil
+		}
+		return nil, err
+	}
+	f, err := bzl.Parse(newfile.Path, b)
+	if err != nil {
+		return nil, err
+	}
+
+	var newStmt []bzl.Expr
+	for _, s := range newfile.Stmt {
+		c, ok := s.(*bzl.CallExpr)
+		if !ok {
+			return nil, fmt.Errorf("got %v expected only CallExpr in %q", s, newfile.Path)
+		}
+		other, err := match(f, c)
+		if err != nil {
+			return nil, err
+		}
+		if other == nil {
+			newStmt = append(newStmt, c)
+			continue
+		}
+		if literal(c.X) == "load" {
+			mergeLoad(c, other, f)
+		} else {
+			merge(c, other)
+		}
+	}
+	f.Stmt = append(f.Stmt, newStmt...)
+	return f, nil
+}
+
+// merge takes new info from src and merges into dest.
+// pre: these calls are the same X and 'name'
+func merge(src, dest *bzl.CallExpr) {
+	for _, item := range src.List {
+		nkv, ok := item.(*bzl.BinaryExpr)
+		if !ok {
+			continue
+		}
+		k := literal(nkv.X)
+		if !mergeableFields[k] {
+			continue
+		}
+		// TODO(pmbethe09): allow '# keep' on src files.
+		replace(dest, nkv)
+	}
+}
+
+func replace(dest *bzl.CallExpr, update *bzl.BinaryExpr) {
+	k := literal(update.X)
+	for i, item := range dest.List {
+		okv, ok := item.(*bzl.BinaryExpr)
+		if !ok {
+			continue
+		}
+		if k == literal(okv.X) {
+			dest.List[i] = update
+			return
+		}
+	}
+	dest.List = append(dest.List, update)
+}
+
+func mergeLoad(src, dest *bzl.CallExpr, oldfile *bzl.File) {
+	vals := make(map[string]bzl.Expr)
+	for _, v := range src.List[1:] {
+		vals[stringValue(v)] = v
+	}
+	for _, v := range dest.List[1:] {
+		rule := stringValue(v)
+		if _, ok := vals[rule]; !ok && ruleUsed(rule, oldfile) {
+			vals[rule] = v
+		}
+	}
+	keys := make([]string, 0, len(vals))
+	for k := range vals {
+		keys = append(keys, k)
+	}
+	sort.Strings(keys)
+	dest.List = dest.List[:1]
+	for _, k := range keys {
+		dest.List = append(dest.List, vals[k])
+	}
+}
+
+func ruleUsed(rule string, oldfile *bzl.File) bool {
+	for _, s := range oldfile.Stmt {
```

Comment:
`len(oldfile.Rules(rule)) != 0` might be simpler, although it is less efficient.